README.md
<h1 align="center">
  TorchFSDD
</h1>

<p align="center">
  <em>A utility for wrapping the Free Spoken Digit Dataset into PyTorch-ready data set splits.</em>
</p>

<p align="center">
  <a href="https://pypi.org/project/torchfsdd">
    <img src="https://img.shields.io/pypi/v/torchfsdd?style=flat-square" alt="PyPI"/>
  </a>
  <a href="https://pypi.org/project/torchfsdd">
    <img src="https://img.shields.io/pypi/pyversions/torchfsdd?style=flat-square" alt="PyPI - Python Version"/>
  </a>
  <a href="https://raw.githubusercontent.com/eonu/torchfsdd/master/LICENSE">
    <img src="https://img.shields.io/pypi/l/torchfsdd?style=flat-square" alt="PyPI - License"/>
  </a>
  <a href="https://torch-fsdd.readthedocs.io/en/latest/">
    <img src="https://readthedocs.org/projects/torch-fsdd/badge/?version=latest&style=flat-square" alt="Read The Docs - Documentation"/>
  </a>
</p>

## About

The [Free Spoken Digit Dataset](https://github.com/Jakobovski/free-spoken-digit-dataset) is an open data set consisting of audio recordings of various individuals speaking the digits from 0-9, with 50 recordings of each digit per individual.

The data set can be thought of as an audio version of the popular [MNIST data set](https://en.wikipedia.org/wiki/MNIST_database) which consists of hand-written digits. However, the fact that the data consists of recordings of different durations makes it more challenging to deal with than the fixed-size images of MNIST.

TorchFSDD aims to provide an interface to FSDD for PyTorch model development, by providing a [`torch.utils.data.Dataset`](https://pytorch.org/docs/stable/data.html#torch.utils.data.Dataset) wrapper that is ready to be used with a [`torch.utils.data.DataLoader`](https://pytorch.org/docs/stable/data.html#torch.utils.data.DataLoader).

## Build status

| `master` | `dev` |
| -------- | ------|
| [![CircleCI Build (Master)](https://img.shields.io/circleci/build/github/eonu/torch-fsdd/master?logo=circleci&style=flat-square)](https://app.circleci.com/pipelines/github/eonu/torch-fsdd?branch=master) | [![CircleCI Build (Development)](https://img.shields.io/circleci/build/github/eonu/torch-fsdd/dev?logo=circleci&style=flat-square)](https://app.circleci.com/pipelines/github/eonu/torch-fsdd?branch=master) |

## Examples

```python
from torchfsdd import TorchFSDDGenerator, TrimSilence
from torchaudio.transforms import MFCC
from torchvision.transforms import Compose

# Create a transformation pipeline to apply to the recordings
transforms = Compose([
    TrimSilence(threshold=1e-6),
    MFCC(sample_rate=8e3, n_mfcc=13)
])

# Fetch the latest version of FSDD and initialize a generator with those files
fsdd = TorchFSDDGenerator(version='master', transforms=transforms)

# Create a Torch dataset for the entire dataset from the generator
full_set = fsdd.full()
# Create two Torch datasets for a train-test split from the generator
train_set, test_set = fsdd.train_test_split(test_size=0.1)
# Create three Torch datasets for a train-validation-test split from the generator
train_set, val_set, test_set = fsdd.train_val_test_split(test_size=0.15, val_size=0.15)
```

A more complete example can be found [here](./notebooks), showing how TorchFSDD can be used to train a neural network.

## Installation

You can install TorchFSDD using `pip`.

```console
pip install torchfsdd
```

**Note**: TorchFSDD assumes you have the following packages already installed (along with Python v3.6+).

- [`torch`](https://github.com/pytorch/pytorch) (>= 1.8.0)
- [`torchaudio`](https://github.com/pytorch/audio) (>= 0.8.0)

Since there are many different possible configurations when installing PyTorch (e.g. CPU or GPU, CUDA version), we leave this up to the user instead of specifying particular versions to install alongside TorchFSDD.

Make sure you have `torch` and `torchaudio` versions that are [compatible](https://github.com/pytorch/audio#dependencies)!

> If you _really_ wish to install `torch` and `torchaudio` together with TorchFSDD automatically, the following will install CPU-only versions of both dependencies.
>
> ```console
> pip install torchfsdd[torch]
> ```

### Development

Please see the [contribution guidelines](/CONTRIBUTING.md) to see installation instructions for contributing to this repository.

## Documentation

Documentation for the package is available on [Read The Docs](https://torch-fsdd.readthedocs.io/en/latest).

## Contributors

All contributions to this repository are greatly appreciated. Contribution guidelines can be found [here](/CONTRIBUTING.md).

<table>
	<thead>
		<tr>
			<th align="center">
        <a href="https://github.com/eonu">
          <img src="https://avatars0.githubusercontent.com/u/24795571?s=460&v=4" alt="eonu" width="60px">
          <br/><sub><b>eonu</b></sub>
        </a>
			</th>
      <th align="center">
        <a href="https://github.com/black-puppydog">
          <img src="https://avatars.githubusercontent.com/u/189241?v=4" alt="black-puppydog" width="60px">
          <br/><sub><b>black-puppydog</b></sub>
        </a>
			</th>
			<!-- Add more <th></th> blocks for more contributors -->
		</tr>
	</thead>
</table>

---

<p align="center">
  <b>TorchFSDD</b> &copy; 2021-2023, Edwin Onuonga - Released under the <a href="https://opensource.org/licenses/MIT">MIT</a> License.<br/>
  <em>Authored and maintained by Edwin Onuonga.</em>
</p>


CHANGELOG.md
# Changelog

## [1.0.0](https://github.com/eonu/torch-fsdd/releases/tag/v1.0.0)

#### Major changes

- Rework `setup.py` to use package extras and fetch metadata from `__init__.py`. ([#15](https://github.com/eonu/torch-fsdd/pull/15))
- Add `dev` package extra for development (including running docs and tests). ([#15](https://github.com/eonu/torch-fsdd/pull/15))

#### Minor changes

- Switch to CircleCI. ([#15](https://github.com/eonu/torch-fsdd/pull/15))
- Add CI for Python versions (3.6 - 3.10). ([#15](https://github.com/eonu/torch-fsdd/pull/15))
- Fix `Jinja2<3.1` version for RTD (see [readthedocs/readthedocs.org/issues/9038](https://github.com/readthedocs/readthedocs.org/issues/9038)). ([#15](https://github.com/eonu/torch-fsdd/pull/15))
- Fix broken ReadTheDocs link — thanks @black-puppydog. ([#13](https://github.com/eonu/torch-fsdd/pull/13))

## [0.1.2](https://github.com/eonu/torch-fsdd/releases/tag/v0.1.2)

#### Major changes

- Don't allow zero `test_size` (see #7). ([#11](https://github.com/eonu/torch-fsdd/pull/11))
- Change `extra_requires` to `extras_require` in `setup.py`. ([#9](https://github.com/eonu/torch-fsdd/pull/9))<br/>(this is why `pip install torchfsdd[torch]` didn't work!)

#### Minor changes

- Bump package development status to beta. ([#10](https://github.com/eonu/torch-fsdd/pull/10))
- Swap `torch` and `torchaudio` links in `README.md`. ([#8](https://github.com/eonu/torch-fsdd/pull/8))

## [0.1.1](https://github.com/eonu/torch-fsdd/releases/tag/v0.1.1)

#### Major changes

- Add `torch` and `torchaudio` dependencies to `docs/requirements.txt`. ([#4](https://github.com/eonu/torch-fsdd/pull/4))
- Upgrade minimum package dependency versions: ([#3](https://github.com/eonu/torch-fsdd/pull/3))
  - `torch` (>= 1.8)
  - `torchaudio` (>= 0.8)
  - `torchvision` (>= 0.9) - only used in tests and notebooks

#### Minor changes

- Remove platform-dependent `torch`, `torchaudio` and `torchvision` installation for tests. ([#1](https://github.com/eonu/torch-fsdd/pull/1))
- Move Travis badges to separate `README.md` section. ([#2](https://github.com/eonu/torch-fsdd/pull/2))

## [0.1.0](https://github.com/eonu/torch-fsdd/releases/tag/v0.1.0)

#### Major changes

Nothing, initial release!

lib/test/test_dataset.py
import os, shutil, pytest, glob, torch
from torchaudio.transforms import MFCC
from torchvision.transforms import Compose
from torchfsdd import TorchFSDD, TorchFSDDGenerator, TrimSilence

def fetch_rec_num(file):
    name = os.path.splitext(os.path.basename(file))[0]
    rec_num = int(name.split('_')[-1])
    return rec_num

def test_generator_local():
    path = 'lib/test/data/v1.0.10'
    fsdd = TorchFSDDGenerator(version='local', path=path)
    with open('lib/test/filenames/v1.0.10') as f:
        true_files = [l.rstrip() for l in f.readlines()]
    read_files = [os.path.basename(file) for file in fsdd.all_files]
    assert len(true_files) == len(read_files) and sorted(true_files) == sorted(read_files)
    assert fsdd.path == path

def test_generator_repo_master():
    path = 'lib/test/data/recordings'
    try:
        fsdd = TorchFSDDGenerator(version='master', path='lib/test/data')
        assert len(fsdd.all_files) > 0
        assert fsdd.path == path
    finally:
        if os.path.isdir(path):
            shutil.rmtree(path)

def test_generator_repo_tag1():
    path = 'lib/test/data/recordings'
    try:
        fsdd = TorchFSDDGenerator(version='v1.0.3', path='lib/test/data')
        with open('lib/test/filenames/v1.0.3') as f:
            true_files = [l.rstrip() for l in f.readlines()]
        read_files = [os.path.basename(file) for file in fsdd.all_files]
        assert len(true_files) == len(read_files) and sorted(true_files) == sorted(read_files)
        assert fsdd.path == path
    finally:
        if os.path.isdir(path):
            shutil.rmtree(path)

def test_generator_repo_tag2():
    path = 'lib/test/data/recordings'
    try:
        fsdd = TorchFSDDGenerator(version='v1.0.10', path='lib/test/data')
        with open('lib/test/filenames/v1.0.10') as f:
            true_files = [l.rstrip() for l in f.readlines()]
        read_files = [os.path.basename(file) for file in fsdd.all_files]
        assert len(true_files) == len(read_files) and sorted(true_files) == sorted(read_files)
        assert fsdd.path == path
    finally:
        if os.path.isdir(path):
            shutil.rmtree(path)

def test_generator_full():
    path = 'lib/test/data/v1.0.10'
    fsdd = TorchFSDDGenerator(version='local', path=path)
    full = fsdd.full()
    with open('lib/test/filenames/v1.0.10') as f:
        true_files = [l.rstrip() for l in f.readlines()]
    read_files = [os.path.basename(file) for file in full.files]
    assert isinstance(full, TorchFSDD)
    assert len(true_files) == len(read_files) and sorted(true_files) == sorted(read_files)
    assert fsdd.path == path

def test_generator_train_test_split_99_1():
    fsdd = TorchFSDDGenerator(version='local', path='lib/test/data/v1.0.10')

    train, test = fsdd.train_test_split(test_size=0.01)
    assert isinstance(train, TorchFSDD)
    assert isinstance(test, TorchFSDD)

    train_nums = {fetch_rec_num(file) for file in train.files}
    assert train_nums == set(range(50))
    assert len(train) == 3000

    test_nums = {fetch_rec_num(file) for file in test.files}
    assert test_nums == set()
    assert len(test) == 0

def test_generator_train_test_split_95_5():
    fsdd = TorchFSDDGenerator(version='local', path='lib/test/data/v1.0.10')

    train, test = fsdd.train_test_split(test_size=0.05)
    assert isinstance(train, TorchFSDD)
    assert isinstance(test, TorchFSDD)

    train_nums = {fetch_rec_num(file) for file in train.files}
    assert train_nums == set(range(2, 50))
    assert len(train) == 2880

    test_nums = {fetch_rec_num(file) for file in test.files}
    assert test_nums == set(range(2))
    assert len(test) == 120

def test_generator_train_test_split_90_10():
    fsdd = TorchFSDDGenerator(version='local', path='lib/test/data/v1.0.10')

    train, test = fsdd.train_test_split(test_size=0.1)
    assert isinstance(train, TorchFSDD)
    assert isinstance(test, TorchFSDD)

    train_nums = {fetch_rec_num(file) for file in train.files}
    assert train_nums == set(range(5, 50))
    assert len(train) == 2700

    test_nums = {fetch_rec_num(file) for file in test.files}
    assert test_nums == set(range(5))
    assert len(test) == 300

def test_generator_train_test_split_50_50():
    fsdd = TorchFSDDGenerator(version='local', path='lib/test/data/v1.0.10')

    train, test = fsdd.train_test_split(test_size=0.5)
    assert isinstance(train, TorchFSDD)
    assert isinstance(test, TorchFSDD)

    train_nums = {fetch_rec_num(file) for file in train.files}
    assert train_nums == set(range(25, 50))
    assert len(train) == 1500

    test_nums = {fetch_rec_num(file) for file in test.files}
    assert test_nums == set(range(0, 25))
    assert len(test) == 1500

def test_generator_train_test_split_10_90():
    fsdd = TorchFSDDGenerator(version='local', path='lib/test/data/v1.0.10')

    train, test = fsdd.train_test_split(test_size=0.9)
    assert isinstance(train, TorchFSDD)
    assert isinstance(test, TorchFSDD)

    train_nums = {fetch_rec_num(file) for file in train.files}
    assert train_nums == set(range(45, 50))
    assert len(train) == 300

    test_nums = {fetch_rec_num(file) for file in test.files}
    assert test_nums == set(range(45))
    assert len(test) == 2700

def test_generator_train_test_split_5_95():
    fsdd = TorchFSDDGenerator(version='local', path='lib/test/data/v1.0.10')

    train, test = fsdd.train_test_split(test_size=0.95)
    assert isinstance(train, TorchFSDD)
    assert isinstance(test, TorchFSDD)

    train_nums = {fetch_rec_num(file) for file in train.files}
    assert train_nums == set(range(47, 50))
    assert len(train) == 180

    test_nums = {fetch_rec_num(file) for file in test.files}
    assert test_nums == set(range(47))
    assert len(test) == 2820

def test_generator_train_test_split_1_99():
    fsdd = TorchFSDDGenerator(version='local', path='lib/test/data/v1.0.10')

    train, test = fsdd.train_test_split(test_size=0.99)
    assert isinstance(train, TorchFSDD)
    assert isinstance(test, TorchFSDD)

    train_nums = {fetch_rec_num(file) for file in train.files}
    assert train_nums == set(range(49, 50))
    assert len(train) == 60

    test_nums = {fetch_rec_num(file) for file in test.files}
    assert test_nums == set(range(49))
    assert len(test) == 2940

def test_generator_train_val_test_split_98_1_1():
    fsdd = TorchFSDDGenerator(version='local', path='lib/test/data/v1.0.10')

    train, val, test = fsdd.train_val_test_split(test_size=0.01, val_size=0.01)
    assert isinstance(train, TorchFSDD)
    assert isinstance(val, TorchFSDD)
    assert isinstance(test, TorchFSDD)

    train_nums = {fetch_rec_num(file) for file in train.files}
    assert train_nums == set(range(50))
    assert len(train) == 3000

    val_nums = {fetch_rec_num(file) for file in val.files}
    assert val_nums == set()
    assert len(val) == 0

    test_nums = {fetch_rec_num(file) for file in test.files}
    assert test_nums == set()
    assert len(test) == 0

def test_generator_train_val_test_split_95_2p5_2p5():
    fsdd = TorchFSDDGenerator(version='local', path='lib/test/data/v1.0.10')

    train, val, test = fsdd.train_val_test_split(test_size=0.025, val_size=0.025)
    assert isinstance(train, TorchFSDD)
    assert isinstance(val, TorchFSDD)
    assert isinstance(test, TorchFSDD)

    train_nums = {fetch_rec_num(file) for file in train.files}
    assert train_nums == set(range(2, 50))
    assert len(train) == 2880

    val_nums = {fetch_rec_num(file) for file in val.files}
    assert val_nums == set(range(1, 2))
    assert len(val) == 60

    test_nums = {fetch_rec_num(file) for file in test.files}
    assert test_nums == set(range(1))
    assert len(test) == 60

def test_generator_train_val_test_split_50_33_17():
    fsdd = TorchFSDDGenerator(version='local', path='lib/test/data/v1.0.10')

    train, val, test = fsdd.train_val_test_split(test_size=0.17, val_size=0.33)
    assert isinstance(train, TorchFSDD)
    assert isinstance(val, TorchFSDD)
    assert isinstance(test, TorchFSDD)

    train_nums = {fetch_rec_num(file) for file in train.files}
    assert train_nums == set(range(24, 50))
    assert len(train) == 1560

    val_nums = {fetch_rec_num(file) for file in val.files}
    assert val_nums == set(range(8, 24))
    assert len(val) == 960

    test_nums = {fetch_rec_num(file) for file in test.files}
    assert test_nums == set(range(8))
    assert len(test) == 480

def test_generator_train_val_test_split_17_50_33():
    fsdd = TorchFSDDGenerator(version='local', path='lib/test/data/v1.0.10')

    train, val, test = fsdd.train_val_test_split(test_size=0.33, val_size=0.50)
    assert isinstance(train, TorchFSDD)
    assert isinstance(val, TorchFSDD)
    assert isinstance(test, TorchFSDD)

    train_nums = {fetch_rec_num(file) for file in train.files}
    assert train_nums == set(range(41, 50))
    assert len(train) == 540

    val_nums = {fetch_rec_num(file) for file in val.files}
    assert val_nums == set(range(16, 41))
    assert len(val) == 1500

    test_nums = {fetch_rec_num(file) for file in test.files}
    assert test_nums == set(range(16))
    assert len(test) == 960

def test_dataset_no_transforms():
    fsdd = TorchFSDD(glob.glob('lib/test/data/v1.0.10/*.wav'))
    x, y = fsdd[0]
    assert isinstance(x, torch.Tensor)
    assert x.ndim == 1
    assert x.min() >= -1
    assert x.max() <= 1
    # assert y == 7

def test_dataset_transforms_single():
    files = glob.glob('lib/test/data/v1.0.10/*.wav')
    x_original, y = TorchFSDD(files)[0]
    x_trans, _ = TorchFSDD(files, transforms=TrimSilence(threshold=0.1))[0]
    assert isinstance(x_trans, torch.Tensor)
    assert x_trans.ndim == 1
    assert x_trans.min() >= -1
    assert x_trans.max() <= 1
    assert len(x_original) != len(x_trans)
    # assert y == 7

def test_dataset_transforms_multiple():
    n_mfcc = 13
    files = glob.glob('lib/test/data/v1.0.10/*.wav')
    x_original, y = TorchFSDD(files)[0]
    x_trans, _ = TorchFSDD(files, transforms=Compose([
        TrimSilence(threshold=0.05),
        MFCC(sample_rate=8e3, n_mfcc=n_mfcc)
    ]))[0]
    assert isinstance(x_trans, torch.Tensor)
    assert x_trans.ndim == 2
    assert len(x_trans) == n_mfcc
    # assert y == 7

def test_dataset_load_all_false():
    files = glob.glob('lib/test/data/v1.0.10/*.wav')
    fsdd = TorchFSDD(files, load_all=False)
    assert not hasattr(fsdd, 'recordings')
    assert not hasattr(fsdd, 'labels')

def test_dataset_load_all_true():
    files = glob.glob('lib/test/data/v1.0.10/*.wav')
    fsdd = TorchFSDD(files, load_all=True)
    assert hasattr(fsdd, 'recordings')
    assert isinstance(fsdd.recordings, list)
    assert isinstance(fsdd.recordings[0], torch.Tensor)
    assert len(fsdd.recordings) == len(files)
    assert hasattr(fsdd, 'labels')
    assert isinstance(fsdd.labels, list)
    assert isinstance(fsdd.labels[0], int)
    assert len(fsdd.labels) == len(files)


lib/test/test_helpers.py
import torch
from torchaudio import load
from torchfsdd import TrimSilence

original, sr = load('lib/test/data/sample.wav')
original = original.flatten()

def test_max_threshold():
    trimmed = TrimSilence(threshold=1.)(original)
    assert len(trimmed) == 0

def test_min_threshold():
    trimmed = TrimSilence(threshold=0.)(original)
    assert torch.eq(trimmed, original).all()

def test_normal_threshold():
    trimmed = TrimSilence(threshold=0.1)(original)
    assert len(trimmed) != len(original)
    assert torch.eq(trimmed, original[9:2301]).all()

notebooks/model.py
import torch
from torch import nn
from torch.nn import functional as F
from torch.nn.utils.rnn import pack_padded_sequence, pad_packed_sequence

class DeepGRU(nn.Module):
    def __init__(self, n_features, n_classes, dims={'gru1': 512, 'gru2': 256, 'gru3': 128, 'fc': 256}):
        super().__init__()
        self.model = nn.ModuleDict({
            'enc': _EncoderNetwork(dims={'in': n_features, **{k:v for k, v in dims.items() if k.startswith('gru')}}),
            'attn': _AttentionModule(dims={'in': dims['gru3']}),
            'clf': _Classifier(dims={'in': dims['gru3']*2, 'fc': dims['fc'], 'out': n_classes}),
        })

    def forward(self, x, x_lengths):
        h, h_last = self.model['enc'](x, x_lengths)
        o_attn = self.model['attn'](h, h_last)
        return self.model['clf'](o_attn)

class _EncoderNetwork(nn.Module):
    def __init__(self, dims):
        super().__init__()
        self.dims = dims
        self.model = nn.ModuleDict({
           'gru1': nn.GRU(self.dims['in'], self.dims['gru1'], num_layers=2, batch_first=True),
           'gru2': nn.GRU(self.dims['gru1'], self.dims['gru2'], num_layers=2, batch_first=True),
           'gru3': nn.GRU(self.dims['gru2'], self.dims['gru3'], num_layers=1, batch_first=True)
        })

    def forward(self, x, x_lengths):
        # Pack the padded Tensor into a PackedSequence
        x_packed = pack_padded_sequence(x, x_lengths, batch_first=True)

        # Pass the PackedSequence through the GRUs
        h_packed, _ = self.model['gru1'](x_packed)
        h_packed, _ = self.model['gru2'](h_packed)
        h_packed, h_last = self.model['gru3'](h_packed)

        # Unpack the hidden state PackedSequence into a padded Tensor
        h_padded = pad_packed_sequence(h_packed, batch_first=True, padding_value=0.0, total_length=max(x_lengths))
        return h_padded[0], h_last
        # Shape: B x T_max x D_out, 1 x B x D_out

class _AttentionModule(nn.Module):
    def __init__(self, dims):
        super().__init__()
        self.dims = dims

        # Attentional context vector weights
        self.W_c = nn.Linear(self.dims['in'], self.dims['in'], bias=False)

        # Auxilliary context
        self.attn_gru = nn.GRU(input_size=self.dims['in'], hidden_size=self.dims['in'])

    def forward(self, h, h_last):
        h_last.transpose_(1, 0)
        # Shape: B x 1 x D_out

        # Calculate attentional context
        h.transpose_(1, 2)
        c = F.softmax(self.W_c(h_last) @ h, dim=0)
        c = (c @ h.transpose(2, 1)).transpose(1, 0)
        # Shape: 1 x B x D_out

        # Calculate auxilliary context
        c_aux, _ = self.attn_gru(c, h_last.transpose(1, 0))
        # Shape: 1 x B x D_out

        # Combine attentional and auxilliary context
        return torch.cat((c.squeeze(0), c_aux.squeeze(0)), dim=1)
        # Shape: B x D_out*2

class _Classifier(nn.Module):
    def __init__(self, dims):
        super().__init__()
        self.dims = dims
        self.model = nn.ModuleDict({
            'fc1': nn.Sequential(
                nn.BatchNorm1d(self.dims['in']),
                nn.Dropout(),
                nn.Linear(self.dims['in'], self.dims['fc'])
            ),
            'fc2': nn.Sequential(
                nn.BatchNorm1d(self.dims['fc']),
                nn.Dropout(),
                nn.Linear(self.dims['fc'], self.dims['out'])
            )
        })

    def forward(self, o_attn):
        f1 = self.model['fc1'](o_attn)
        f2 = self.model['fc2'](F.relu(f1))
        return F.log_softmax(f2, dim=1)

lib/torchfsdd/dataset.py
import os, shutil, subprocess, glob, torch, torchaudio

REPOSITORY = {
    'name': 'free-spoken-digit-dataset',
    'url': 'https://github.com/Jakobovski/free-spoken-digit-dataset'
}

N_REC = 50

class TorchFSDDGenerator:
    """A :class:`torch:torch.utils.data.Dataset` generator for splits of the Free Spoken Digit Dataset.

    Parameters
    ----------
    version: str
        The version of FSDD to download from `the GitHub repository <https://github.com/Jakobovski/free-spoken-digit-dataset>`_,
        specified as a branch name (defaults to `'master'`) or Git version tag, e.g. `'v1.0.6'`.

        Alternatively, if you already have a local copy of the dataset that you would like to use,
        you can set this argument to `'local'` and provide a path to the folder containing the WAV files, as the ``path`` argument.

    path: str, optional
        If ``version`` is a Git branch name or version tag, then this is the path where the Git repository will be cloned to
        (a new folder will be created at the specified path). If none is specified, then :py:func:`python:os.getcwd` is used.

        If ``version`` is set to `'local'`, then this is the path to the folder containing the WAV audio recordings.

    transforms: callable, optional
        A callable transformation to apply to a 1D :class:`torch:torch.Tensor` of audio samples.

        .. seealso::

            :class:`TorchFSDD`

    load_all: bool
        Whether or not to load the entire dataset into memory.

        .. seealso::

            :class:`TorchFSDD`

    **args: optional
        Arbitrary keyword arguments passed on to :py:func:`torchaudio:torchaudio.load`.
    """
    def __init__(self, version='master', path=None, transforms=None, load_all=False, **args):
        if version == 'local':
            if path is None:
                raise ValueError('Expected path to be a directory containing WAV recordings')
        else:
            path = os.getcwd() if path is None else path
            repo_path = os.path.join(path, REPOSITORY['name'])

            try:
                subprocess.call(['git', '-C', path, 'clone', REPOSITORY['url'], '--branch', version])
                shutil.move(os.path.join(repo_path, 'recordings'), path)
                path = os.path.join(path, 'recordings')
            finally:
                if os.path.isdir(repo_path):
                    shutil.rmtree(repo_path)

        self.path = path
        self.transforms = transforms
        self.load_all = load_all
        self.args = args
        self.all_files = glob.glob(os.path.join(self.path, '*.wav'))

    def full(self):
        """Generates a data set wrapper for the entire data set.

        Returns
        -------
        full_set: :class:`TorchFSDD`
            The :class:`torch:torch.utils.data.Dataset` wrapper for the full data set.
        """
        return TorchFSDD(self.all_files, self.transforms, self.load_all, **self.args)

    def train_test_split(self, test_size=0.1):
        """Generates training and test data set wrappers.

        Parameters
        ----------
        test_size: 0 < float < 1
            Size of the test data set (as a proportion).

        Returns
        -------
        train_set: :class:`TorchFSDD`
            The training set :class:`torch:torch.utils.data.Dataset` wrapper.

        test_set: :class:`TorchFSDD`
            The test set :class:`torch:torch.utils.data.Dataset` wrapper.
        """
        assert 0. < test_size < 1.

        train_files, test_files = [], []
        n_test = int(N_REC * test_size)

        for file in self.all_files:
            file_name, ext = os.path.splitext(os.path.basename(file))
            digit, name, rec_num = file_name.split('_')
            split = test_files if int(rec_num) + 1 <= n_test else train_files
            split.append(file)

        train_set = TorchFSDD(train_files, self.transforms, self.load_all, **self.args)
        test_set = TorchFSDD(test_files, self.transforms, self.load_all, **self.args)
        return train_set, test_set

    def train_val_test_split(self, test_size=0.1, val_size=0.1):
        """Generates training, validation and test data set wrappers.

        Parameters
        ----------
        test_size: 0 < float < 1
            Size of the test data set (as a proportion).

        val_size: 0 < float < 1
            Size of the validation data set (as a proportion).

        Returns
        -------
        train_set: :class:`TorchFSDD`
            The training set :class:`torch:torch.utils.data.Dataset` wrapper.

        val_set: :class:`TorchFSDD`
            The validation set :class:`torch:torch.utils.data.Dataset` wrapper.

        test_set: :class:`TorchFSDD`
            The test set :class:`torch:torch.utils.data.Dataset` wrapper.
        """
        assert 0. < test_size < 1.
        assert 0. < val_size < 1.
        assert test_size + val_size < 1.

        train_files, val_files, test_files = [], [], []
        n_test, n_val = int(N_REC * test_size), int(N_REC * val_size)

        for file in self.all_files:
            file_name, ext = os.path.splitext(os.path.basename(file))
            digit, name, rec_num = file_name.split('_')
            rec_num = int(rec_num) + 1
            if rec_num <= n_test:
                split = test_files
            elif n_test < rec_num <= n_test + n_val:
                split = val_files
            else:
                split = train_files
            split.append(file)

        train_set = TorchFSDD(train_files, self.transforms, self.load_all, **self.args)
        val_set = TorchFSDD(val_files, self.transforms, self.load_all, **self.args)
        test_set = TorchFSDD(test_files, self.transforms, self.load_all, **self.args)
        return train_set, val_set, test_set

class TorchFSDD(torch.utils.data.Dataset):
    """A :class:`torch:torch.utils.data.Dataset` wrapper for specified
    WAV audio recordings of the Free Spoken Digit Dataset.

    .. tip::

        There should rarely be a situation where you have to initialize this class manually,
        unless you are experimenting with specific subsets of the FSDD. You should use :class:`TorchFSDDGenerator`
        to either load the full data set or generate splits for training/validation/testing.

    Parameters
    ----------
    files: list of str
        List of file paths to the WAV audio recordings for the dataset.

    transforms: callable, optional
        A callable transformation to apply to a 1D :class:`torch:torch.Tensor` of audio samples.

        This can be a single transformation, such as the :class:`TrimSilence` transformation included in this package.

        .. code-block:: python

            from torchfsdd import TorchFSDDGenerator, TrimSilence

            fsdd = TorchFSDDGenerator(transforms=TrimSilence(threshold=150))

        It could also be a series of transformations composed together with :class:`torchvision:torchvision.transforms.Compose`.

        .. code-block:: python

            from torchfsdd import TorchFSDDGenerator, TrimSilence
            from torchaudio.transforms import MFCC
            from torchvision.transforms import Compose

            fsdd = TorchFSDDGenerator(transforms=Compose([
                TrimSilence(threshold=100),
                MFCC(sample_rate=8e3, n_mfcc=13)
            ]))

        There are many useful audio transformations in :py:mod:`torchaudio:torchaudio.transforms` such as :class:`torchaudio:torchaudio.transforms.MFCC`.

    load_all: bool
        Whether or not to load the entire dataset into memory.

        This essentially defeats the point of batching, but the dataset is relatively small
        enough that it can comfortably fit into memory and possibly provide some speed-up.

        If this is set to `True`, then the complete set of raw audio recordings and labels
        (for the specified files) can be accessed with ``self.recordings`` and ``self.labels``.

    **args: optional
        Arbitrary keyword arguments passed on to :py:func:`torchaudio:torchaudio.load`.
    """
    def __init__(self, files, transforms=None, load_all=False, **args):
        super().__init__()
        self.files = files
        self.transforms = transforms
        self.args = args

        get_audio = lambda file: torchaudio.load(file, **self.args)[0]
        get_label = lambda file: int(os.path.basename(file)[0])

        if load_all:
            self.recordings, self.labels = [], []
            for file in self.files:
                self.recordings.append(get_audio(file))
                self.labels.append(get_label(file))

            def _load(self, index):
                return self.recordings[index], self.labels[index]
        else:
            def _load(self, index):
                file = self.files[index]
                return get_audio(file), get_label(file)

        setattr(self.__class__, '_load', _load)

    def __len__(self):
        return len(self.files)

    def __getitem__(self, index):
        # Fetch the audio and corresponding label
        x, y = self._load(index)
        x = x.flatten()

        # Transform data if a transformation is given
        if self.transforms is not None:
            x = self.transforms(x)

        return x, y

lib/torchfsdd/helpers.py
class TrimSilence:
    """Removes the silence at the beginning and end of the passed audio data.

    .. warning::
        This transformation assumes that the audio **is** normalized.

    Parameters
    ----------
    threshold: float
        The maximum amount of noise that is considered silence.
    """
    def __init__(self, threshold):
        assert 0. <= threshold <= 1.
        self.threshold = threshold

    def __call__(self, x):
        """Applies the transformation.

        Parameters
        ----------
        x: torch.Tensor
            A one-dimensional tensor of WAV audio samples.

        Returns
        -------
        x: :class:`torch:torch.Tensor`
            The original tensor trimmed for silence.
        """
        start, end = 0, 0

        for i, sample in enumerate(x):
            if abs(sample) > self.threshold:
                start = i
                break

        # Reverse the array for trimming the end
        for i, sample in enumerate(x.flip(dims=(0,))):
            if abs(sample) > self.threshold:
                end = len(x) - i
                break

        return x[start:end]

lib/torchfsdd/__init__.py
import importlib, platform
from pkg_resources import packaging
if packaging.version.parse(platform.python_version()) < packaging.version.parse('3.8'):
    import importlib_metadata as metadata
else:
    from importlib import metadata

__name__ = "torchfsdd"
__version__ = "1.0.0"
__author__ = "Edwin Onuonga"
__email__ = "ed@eonu.net"
__copyright__ = f"2019-2023, {__author__}"

MIN_TORCH_VERSION = '1.8'
MIN_TORCHAUDIO_VERSION = '0.8'

def check_package(pkg, min_version, url):
    """Checks whether a specified package has been installed,
    and whether the installed version meets a specified minimum.

    Parameters
    ----------
    pkg: str
        Name of the package.

    min_version: str
        Minimum version for the package, e.g. `1.8`.

    url: str
        Package installation page URL (for help).
    """
    try:
        importlib.import_module(pkg)
    except ImportError:
        msg = ("Could not find a valid installation of '{pkg}' (>={min_version}), which TorchFSDD depends on.\n"
        "Visit {url} for more instructions on installing this package.").format(pkg=pkg, url=url, min_version=min_version)
        raise ModuleNotFoundError(msg)

    installed_version = metadata.version(pkg)
    if packaging.version.parse(installed_version) < packaging.version.parse(min_version):
        msg = ("Could not find a compatible installation of '{pkg}' (>={min_version}), which TorchFSDD depends on - got version {installed_version}.\n"
        "Visit {url} for more instructions on installing this package.").format(pkg=pkg, url=url, min_version=min_version, installed_version=installed_version)
        raise ImportWarning(msg)

# Check that the minimum dependency versions are installed
check_package('torch', MIN_TORCH_VERSION, url='https://pytorch.org/')
check_package('torchaudio', MIN_TORCHAUDIO_VERSION, url='https://github.com/pytorch/audio')

# Import classes from the package
from .dataset import TorchFSDD, TorchFSDDGenerator
from .helpers import TrimSilence

__all__ = ['TorchFSDD', 'TorchFSDDGenerator', 'TrimSilence']
